microprofile/WFLY-12313_MicroProfile_OpenAPI_1.1.adoc

```diff
@@ -0,0 +1,126 @@
+** Which feature packs will use this layer is still TBD.
+* The microprofile-openapi spec module must be exposed to web deployments
+* The `/openapi` endpoint shall be registered relative to the root of the virtual host associated with the deployment.
+** N.B. The `/openapi` endpoint is intended for a public interface, not a management interface, as with the `/health` and `/metrics` endpoints.
+** This places a superficial constraint of a single application per host (see subsequent section re: multiple deployments)
+** In the event a web deployment uses a given host for which another deployment has already registered an `/openapi` endpoint, we will:
+... Log a warning.
+... Skip OpenAPI model generation for that deployment.
+* All WildFly/SmallRye specific configuration will be obtained via MicroProfile Config using the properties prefixed with `mp.openapi.extensions`.
+** Consequently, the microprofile-openapi subsystem should have an empty management model.
+* Support YAML and JSON for both input (static/pre-generated OpenAPI document) and output formats.
+* MicroProfile OpenAPI TCK must pass
+
+=== Nice-to-Have Requirements
+
+* Optional features of the specification:
+** Multiple deployments:
+*** If a container allows multiple deployments, the specification expects the `/openapi` endpoint to return the logical union of the documented services for all applications in the runtime. See: https://download.eclipse.org/microprofile/microprofile-open-api-1.1.2/microprofile-openapi-spec.html#_multiple_applications
+**** Initially, we will limit multiple deployment support to a single deployment per host (for a given endpoint).
+**** Merging (and more interestingly, unmerging on undeploy) OpenAPI models for all applications deployed to a given host.
+***** Consider for a future enhancement
+** HTTPS transport for the `/openapi` endpoint.
+*** If the server/host of a given application defines an HTTPS listener, we will allow it, but not require it.
+*** Additionally, if the Undertow server/host associated with the application does not define an HTTP endpoint (as required by the spec), we should log a WARN message.
+** The MicroProfile OpenAPI specification suggests, but does not require, the use of a `format=JSON` or `format=YAML` parameter for the `/openapi` endpoint, used to define content type of the resulting document, in lieu of an `Accept` header.
+*** This is very reasonable, and is simple enough to do.  See: https://download.eclipse.org/microprofile/microprofile-open-api-1.1.2/microprofile-openapi-spec.html#_query_parameters
+* Use a `mp.openapi.extensions.disable` property to disable OpenAPI documentation globally or per-application.
+** Provides a parameterizable means of disabling OpenAPI documentation in specific environments (e.g. prod vs dev)
+** Provides a mechanism to control which application associated with a given virtual host should generate an MicroProfile OpenAPI model.
```

Comment:
> Provides a mechanism...

From the user point of view, given the [supposed] logic behind the property name, I take that this one would be a per-exclusion and  derivative mechanism, correct?

Reply:
Yes - the property would be applied according to the ConfigSource scope and its ordinal.

Reply:
As said previously, from user point of view and in order to understand the logic in detail, IMO more information about how the implementation will prioritize global vs application with respect to MP OpenAPI module configuratio properties should be added here.

Reply:
> how the implementation will prioritize global vs application with respect to MP OpenAPI module configuratio properties

We don't control this.  The prioritization of properties is determined according to the MP Config spec.

Reply:
Right I'll have a look to see whether the reader can derive the information from there and will get back.

Reply:
I went back to MP Config specs and - apart from ordinal definition for config sources - no great detail is shared on how implementation should deal with configuration properties/sources priority.

Reply:
That is all there is to it.  The ordinal priority of a ConfigSource determines which configuration property values take precedence.

Reply:
Ok, thanks for confirmation.